pkg/backup/backuper.go

```diff
@@ -0,0 +1,356 @@
+package backup
+
+import (
+	"context"
+	"database/sql"
+	"fmt"
+	"os"
+	"path"
+	"time"
+
+	"github.com/mattn/go-sqlite3"
+
+	"github.com/pkg/errors"
+)
+
+// BackupFilenamePrefix is the prefix used in every backup file.
+const BackupFilenamePrefix = "tbl_backup"
+
+// Backuper is the process that executes the backup process.
+type Backuper struct {
+	sourcePath, dir string
+	source          DB
+	backup          DB
+	config          *Config
+
+	fileCreator func(string, time.Time) (string, error)
+}
+
+// NewBackuper creates a new backuper responsible for making backups of a SQLite database.
+func NewBackuper(sourcePath string, backupDir string, opts ...Option) (*Backuper, error) {
+	config := DefaultConfig()
+	for _, o := range opts {
+		if err := o(config); err != nil {
+			return nil, err
+		}
+	}
+
+	if err := os.MkdirAll(backupDir, 0o755); err != nil {
+		return nil, errors.Errorf("os mkdir all: %s", err)
+	}
+
+	b := &Backuper{
+		sourcePath:  sourcePath,
+		dir:         backupDir,
+		config:      config,
+		fileCreator: createBackupFile,
+	}
+
+	return b, nil
+}
+
+// Init opens databases and ping them, then initializes variables.
+func (b *Backuper) Init() error {
+	source, err := open(b.sourcePath)
+	if err != nil {
+		return errors.Errorf("opening source db: %s", err)
+	}
+
+	timestamp := time.Now().UTC()
+	filename, err := b.fileCreator(b.dir, timestamp)
+	if err != nil {
+		return errors.Errorf("creating backup file: %s", err)
+	}
+
+	backup, err := open(filename)
+	if err != nil {
+		return errors.Errorf("opening backup db: %s", err)
+	}
+
+	b.source = source
+	b.backup = backup
+	return nil
+}
+
+// Backup creates a backup to a file in disk.
+// Multiple calls to Backup can be perfomed.
+// But the file in which backup occurs was already created, which means the backup always occur at the same file.
+// This can be used to perform retries in case of errors.
+func (b *Backuper) Backup(ctx context.Context) (_ BackupResult, err error) {
+	defer func() {
+		if err != nil {
```

Comment:
If any error occurs in the backup process we'll delete the temporary backup file